### PR TITLE
Fix isp violations

### PR DESCRIPTION
**Pull request by greg0ire**
I am targeting this branch, because this is BC.

## Subject

The admin interface should be split into smaller interfaces, as per the ISP. This should ease the splitting in of the `AbstractAdmin` class itself.

## Changelog
```markdown
### Added
- `Sonata\AdminBundle\Admin\AdminInterface` was split into smaller interfaces, namely
    - `Sonata\AdminBundle\AdminFieldDescriptionRegistryInterface`
    - `Sonata\AdminBundle\LifecycleHookProviderInterface`
    - `Sonata\AdminBundle\MenuBuilderInterface`
    - `Sonata\AdminBundle\ParentAdminInterface`
```

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Why this should be pedantic? It's code modification, it should be at least a patch.

Plus, you introduce new interfaces: Maybe a minor? Some users may rely on them.

**Comment by greg0ire** [comment]
I'm not fixing anything apart from a solid principle violation. Minor might be appropriate though.

File: Admin/FieldDescriptionRegistryInterface.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+     * Returns true if the filter FieldDescription exists.
+     *
+     * @param string $name
+     *
+     * @return bool
+     */
+    public function hasFilterFieldDescription($name);
+
+    /**
+     * add a filter FieldDescription.
```
Comment (core23) [inline Admin/FieldDescriptionRegistryInterface.php]:
```diff
- add
+ Adds
```

File: Admin/ParentAdminInterface.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+/**
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+interface ParentAdminInterface
```
Comment (core23) [inline Admin/ParentAdminInterface.php]: Please add a description for this (and the other interfaces)

**Comment by core23** [comment]
👍 for a minor change, because we introduce new classes / features

File: Admin/AdminInterface.php [hunk 3]
```diff
@@ -145,7 +143,7 @@ public function getBaseControllerName();
     public function generateObjectUrl($name, $object, array $parameters = array(), $absolute = false);
 
     /**
-     * Generates an url for the given parameters.
+     * Generates a url for the given parameters.
```
Comment (core23) [inline Admin/AdminInterface.php]: Why did you change the article? Should be `an`, because a vowel is following.

File: Admin/LifecycleHookProviderInterface.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+interface LifecycleHookProviderInterface
+{
+    /**
+     * @param mixed $object
+     *
+     * @return mixed
```
Comment (core23) [inline Admin/LifecycleHookProviderInterface.php]: Should we use `object` instead of mixed?

File: Admin/MenuBuilderInterface.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+use Knp\Menu\FactoryInterface as MenuFactoryInterface;
+use Knp\Menu\ItemInterface;
+
+/**
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+interface MenuBuilderInterface
+{
+    /**
+     * @param \Knp\Menu\FactoryInterface $menuFactory
```
Comment (core23) [inline Admin/MenuBuilderInterface.php]: No FQCN please

File: Admin/AdminInterface.php (hunk 3 above)
Comment (greg0ire) [inline Admin/AdminInterface.php]: That's not how it works. U starts with a consonant sound. In phonetic, it's spelt "ju", check it out ;)

File: Admin/LifecycleHookProviderInterface.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    public function create($object);
+
+    /**
+     * @param mixed $object
+     */
+    public function delete($object);
+
+//TODO: uncomment this method for 4.0
```
Comment (jlamur) [inline Admin/LifecycleHookProviderInterface.php]: Maybe `NEXT_MAJOR` instead of `TODO` here?

**Comment by greg0ire** [comment]
Removing the WIP prefix, I think this can be merged as is: we can split even further after that if we feel the need.

File: Admin/FieldDescriptionRegistryInterface.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+interface FieldDescriptionRegistryInterface
```
Comment (core23) [inline Admin/FieldDescriptionRegistryInterface.php]: Can you please add a description to the interface

File: Admin/FieldDescriptionRegistryInterface.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+     * @param string $name
+     */
+    public function removeListFieldDescription($name);
+
+    /**
+     * Returns a list depend on the given $object.
+     *
+     * @return FieldDescriptionCollection
+     */
+    public function getList();
```
Comment (core23) [inline Admin/FieldDescriptionRegistryInterface.php]: This method exists in the normal `AdminInterface`? The name of the method sounds very general.

File: Admin/LifecycleHookProviderInterface.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+namespace Sonata\AdminBundle\Admin;
+
+/**
+ * This interface can be implemented to provide hooks that will be called
+ * during the lifecycle of the object.
+ */
+interface LifecycleHookProviderInterface
+{
+    /**
+     * @param object $object
```
Comment (core23) [inline Admin/LifecycleHookProviderInterface.php]: I know that you just moved the methods. But can you find useful descriptions for the methods?

File: Admin/LifecycleHookProviderInterface.php [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+/**
+ * This interface can be implemented to provide hooks that will be called
+ * during the lifecycle of the object.
+ */
+interface LifecycleHookProviderInterface
```
Comment (core23) [inline Admin/LifecycleHookProviderInterface.php]: Maybe add an `@author` tag?

File: Admin/MenuBuilderInterface.php [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+{
+    /**
+     * @param MenuFactoryInterface $menuFactory
+     */
+    public function setMenuFactory(MenuFactoryInterface $menuFactory);
+
+    /**
+     * @return MenuFactoryInterface
+     */
+    public function getMenuFactory();
```
Comment (core23) [inline Admin/MenuBuilderInterface.php]: Don't know if we should add the getter and setter methods to the interface

File: Admin/AdminInterface.php [hunk 12]
```diff
@@ -34,7 +32,7 @@
 /**
  * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
  */
-interface AdminInterface
+interface AdminInterface extends FieldDescriptionRegistryInterface, LifecycleHookProviderInterface, MenuBuilderInterface, ParentAdminInterface
```
Comment (core23) [inline Admin/AdminInterface.php]: Maybe create  an `RouterInterface` for all routing methods like `generate*Url`

**Comment by core23** [comment]
Greate work ❤️ 

Can you find some time to move some more methods to separate interfaces

**Comment by greg0ire** [comment]
> Can you find some time to move some more methods to separate interfaces

I'll try that this afternoon :)

File: Admin/AdminInterface.php (hunk 12 above)
Comment (greg0ire) [inline Admin/AdminInterface.php]: I think it makes sense to put `setRouteGenerator` in there too

Comment (greg0ire) [inline Admin/AdminInterface.php]: As well as `getRoutes`, `getRouterIdParameter`, and `getUrlsafeIdentifier`

File: Admin/AdminInterface.php [hunk 13]
```diff
@@ -327,92 +255,6 @@ public function hasRoute($name);
...
-     * @return FieldDescriptionInterface|null
-     */
-    public function getFilterFieldDescription($name);
-
-    /**
-     * Returns a list depend on the given $object.
-     *
-     * @return FieldDescriptionCollection
-     */
-    public function getList();
```
Comment (greg0ire) [inline Admin/AdminInterface.php]: @core23 yeah this method existed

File: Admin/FieldDescriptionRegistryInterface.php (hunk 8 above)
Comment (greg0ire) [inline Admin/FieldDescriptionRegistryInterface.php]: Yeah, see https://github.com/sonata-project/SonataAdminBundle/pull/4440/files#r128921201

File: Admin/MenuBuilderInterface.php (hunk 11 above)
Comment (greg0ire) [inline Admin/MenuBuilderInterface.php]: Having setter methods is probably not a great idea, but since they are in the original interface, I'm forced to keep them. What are you suggesting exactly? Removing them in a next major version, or moving them to another interface right now?

File: Admin/LifecycleHookProviderInterface.php (hunk 9 above)
Comment (greg0ire) [inline Admin/LifecycleHookProviderInterface.php]: I found things to say about create, update and delete, but have nothing interesting to say about the other methods.

**Comment by greg0ire** [comment]
@core23 I think I addressed all your comments, please review again. @sonata-project/contributors please do so too again.

**Comment by greg0ire** [comment]
I think I can provide one more interface, about access control.

File: Admin/AccessRegistryInterface.php [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * Hook to handle access authorization.
+     *
+     * @param string $action
+     * @param object $object
+     */
+    public function checkAccess($action, $object = null);
+
+    /*
```
Comment (greg0ire) [inline Admin/AccessRegistryInterface.php]: StyleCI forced me to remove a star here, since this is not a real method yet.

File: Admin/FieldDescriptionRegistryInterface.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+     * @param string $name
+     *
+     * @return FieldDescriptionInterface
+     */
+    public function getFormFieldDescription($name);
+
+    /**
+     * Build and return the collection of form FieldDescription.
+     *
+     * @return array collection of form FieldDescription
```
Comment (hhamon) [inline Admin/FieldDescriptionRegistryInterface.php]: `FieldDescriptionInterface[]` ?

Comment (greg0ire) [inline Admin/FieldDescriptionRegistryInterface.php]: @hhamon fixed

File: Admin/MenuBuilderInterface.php (hunk 11 above)
Comment (core23) [inline Admin/MenuBuilderInterface.php]: Maybe keep it in the `AdminInterface` and deprecate the whole `AdminInterface`

**Comment by OskarStark** [comment]
how to proceed here @greg0ire ? I would like to get this merged!

**Comment by greg0ire** [comment]
I'll try to have a look at this soon

**Comment by greg0ire** [comment]
@OskarStark I rebased. @sonata-project/contributors please review again

File: Admin/MenuBuilderInterface.php (hunk 11 above)
Comment (core23) [inline Admin/MenuBuilderInterface.php]: ping @greg0ire

Comment (greg0ire) [inline Admin/MenuBuilderInterface.php]: Good idea

Comment (greg0ire) [inline Admin/MenuBuilderInterface.php]: Oh wait there are way too many setters in `Admin/FieldDescriptionInterface.php`. Should I really move them back? What exact methods would you move back and why?

Comment (greg0ire) [inline Admin/MenuBuilderInterface.php]: Also, how do you deprecate an interface? And is removing methods from an interface a BC-break?

Comment (core23) [inline Admin/MenuBuilderInterface.php]: > Should I really move them back? What exact methods would you move back and why?

Just talking about the `getMenuFactory` method. IMHO this method did only exist for BR reasons.


> Also, how do you deprecate an interface? And is removing methods from an interface a BC-break?

You're right. We must keep with the next major release, because we rely on it in many files.

Comment (greg0ire) [inline Admin/MenuBuilderInterface.php]: What's BR?

Comment (greg0ire) [inline Admin/MenuBuilderInterface.php]: I moved `getMenuFactory` and `setMenuFactory` back to the admin interface. I hope that is what you wanted.

**Comment by greg0ire** [comment]
Blocked by https://github.com/sonata-project/dev-kit/pull/267

**Comment by greg0ire** [comment]
Restarting the build

**Comment by core23** [comment]
@OskarStark Please review again, so we can merge this

**Comment by jordisala1991** [comment]
Nice job @greg0ire ! :)

**Comment by Soullivaneuh** [comment]
@greg0ire could you please review your changelog? It should be small and concise and not multi-line.

Class list should be on the upgrade section instead.
